### PR TITLE
Add a README file

### PR DESCRIPTION
**Pull request by igortg**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# pytest-update-test-results
+
+Pytest plugin to update test results on a test report from previous run.
+
+The primary purpose of this plugin is to avoid flaky tests to break a CI
+build by executing them again after a failed run. The plugin does not
+reruns tests by itself, but allow the user to leverage the pytest
+`--last-failed` option to execute failed tests in a separete run and
```
Comment (lincrosenbach) [inline README.md]: Typo `separete`

File: README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+like beatign a dead horse. Hence, we tried `--reruns` option from **pytest-rerunfailres** (which re-run all test failures), but that didn't
+work due the way **pytest-rerunfailres** [deal with session fixtures](1).
+
+So we opted to develop this plugin to re-run tests in a separate pytest
+execution and update an existing XML report. Besides working just fine with session fixtures, re-running failed tests in a separate process has other advantages:
+
+1. By starting a separate process, we make sure no "test contamination"
+occurs (memory leaks vanish, objects that live during the entire test
+session are reset).
+1. Since we heavily parallelize the test suite (using **pytest-xdist**),
```
Comment (lincrosenbach) [inline README.md]: Should be 2.?

**Review by lincrosenbach — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# pytest-update-test-results
+
+Pytest plugin to update test results on a test report from previous run.
```
Comment (nicoddemus) [inline README.md]:
```suggestion
Pytest plugin to update test results on a test report from a previous run.
```

File: README.md (hunk 1 above)
Comment (nicoddemus) [inline README.md]:
```suggestion
`--last-failed` option to execute failed tests in a separate run and
```

File: README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+build by executing them again after a failed run. The plugin does not
+reruns tests by itself, but allow the user to leverage the pytest
+`--last-failed` option to execute failed tests in a separete run and
+provides `--update-xml` option so if previously failed tests
+passes, the XML test report will be updated (and as result, your CI
+would flag the test as successfull).
+
+In summary, your CI script would have something like:
+
+```bash
```
Comment (nicoddemus) [inline README.md:15]: Nice!

File: README.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    pytest tests/my-test-suite --last-faield --update-xml pytest.xml
+fi
+```
+
+## Alternatives
+
+In our particular cases, most of the flakiness comes from using a
+shared storage on our virtualized on-premise CI cluster, which made for
+random file access times to varies in order of 10x-100x depending on
+the cluster loading. This do not affect small unit tests, but greatly
```
Comment (nicoddemus) [inline README.md]:
```suggestion
the cluster loading. This does not affect small unit tests, but greatly
```

File: README.md (hunk 2 above)
Comment (nicoddemus) [reply README.md]: It is good practice to always use 1, and let the Markdown renderer worry about numbering (makes it easier to reorder/add new sessions at any order later).

File: README.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+affects execution time of integration and end-to-end tests, resulting
+in flaky tests due timeout or race condition.
+
+In this scenario, marking individual tests as flaky (using [flaky]) was
+like beatign a dead horse. Hence, we tried `--reruns` option from **pytest-rerunfailres** (which re-run all test failures), but that didn't
```
Comment (nicoddemus) [inline README.md]:
```suggestion
like beating a dead horse. 

Next, we tried using [pytest-rerunfailures](pytest-rerunfailures) (which re-run all test failures), but that didn't
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+work due the way **pytest-rerunfailres** [deal with session fixtures](1).
```
Comment (nicoddemus) [inline README.md]:
```suggestion
work due the way it [deals with session fixtures](1).
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+So we opted to develop this plugin to re-run tests in a separate pytest
```
Comment (nicoddemus) [inline README.md]:
```suggestion
So we opted to develop this plugin to re-run just the failed tests from the previous run in a new pytest
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+execution and update an existing XML report. Besides working just fine with session fixtures, re-running failed tests in a separate process has other advantages:
```
Comment (nicoddemus) [inline README.md]:
```suggestion
execution and update the XML report. Besides working just fine with session fixtures, re-running failed tests in a separate process has other advantages:
```

File: README.md (hunk 2 above)
Comment (nicoddemus) [inline README.md]:
```suggestion
1. Because we heavily parallelize the test suite (using **pytest-xdist**),
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+1. By starting a separate process, we make sure no "test contamination"
+occurs (memory leaks vanish, objects that live during the entire test
+session are reset).
+1. Since we heavily parallelize the test suite (using **pytest-xdist**),
+some flaky tests can occur due high CPU/IO loading. In the re-run
```
Comment (nicoddemus) [inline README.md]:
```suggestion
some flaky tests can occur due to high CPU/IO usage. In the re-run
```

File: setup.py [hunk 11]
```diff
@@ -22,7 +22,7 @@
         "Programming Language :: Python :: 3.7",
         "Framework :: Pytest",
     ],
-    description="Pytest plugin to update test results on a jUnit XML from previous run.",
+    description="Pytest plugin to update test results on a test report from previous run.",
```
Comment (nicoddemus) [inline setup.py]: I think it is nice to specify the JUnit XML report, any reason you changed this?

**Review by nicoddemus — APPROVED** [review]
Nice!

File: README.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+pytest tests/my-test-suite --junix-xml pytest.xml
+
+if [ $? -eq 2 ]  # pytest exit code for test failure
+then
+    # re-run failed test and update previous run report.
+    pytest tests/my-test-suite --last-faield --update-xml pytest.xml
+fi
+```
+
+## Alternatives
```
Comment (nicoddemus) [inline README.md]: I think this conveys better:

```suggestion
## Rationale
```

As you also explain the adopted solution and the reasons for doing so.

File: README.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+In summary, your CI script would have something like:
+
+```bash
+#/bin/bash
+pytest tests/my-test-suite --junix-xml pytest.xml
+
+if [ $? -eq 2 ]  # pytest exit code for test failure
+then
+    # re-run failed test and update previous run report.
+    pytest tests/my-test-suite --last-faield --update-xml pytest.xml
```
Comment (eduardokern) [inline README.md]:
```suggestion
    pytest tests/my-test-suite --last-failed --update-xml pytest.xml
```

File: README.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+1. By starting a separate process, we make sure no "test contamination"
+occurs (memory leaks vanish, objects that live during the entire test
+session are reset).
+1. Since we heavily parallelize the test suite (using **pytest-xdist**),
+some flaky tests can occur due high CPU/IO loading. In the re-run
+we can reduce number of parallel processes to make more resources
+available for each test.
+
+[flaky]: https://github.com/box/flaky
+[pytest-rerunfailres]: https://github.com/pytest-dev/pytest-rerunfailures
```
Comment (eduardokern) [inline README.md]:
```suggestion
[pytest-rerunfailures]: https://github.com/pytest-dev/pytest-rerunfailures
```

File: README.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
+# pytest-update-test-results
+
+Pytest plugin to update test results on a test report from previous run.
+
+The primary purpose of this plugin is to avoid flaky tests to break a CI
+build by executing them again after a failed run. The plugin does not
+reruns tests by itself, but allow the user to leverage the pytest
```
Comment (LucasRochaAbraao) [inline README.md]:
```suggestion
build by executing them again after a failed run. The plugin does not
rerun tests by itself, but allow the user to leverage the pytest
```

**Review by LucasRochaAbraao — APPROVED** [review]

**Review by eduardokern — APPROVED** [review]

**Review by tadeu — APPROVED** [review]

**Comment by igortg** [comment]
Thanks for the thorough review everyone.

File: setup.py (hunk 11 above)
Comment (igortg) [reply setup.py]: Nop. Better to be explict about it. Fixed.

**Pull request merged**